src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,255 @@
+    QDomElement metadata = emptyTag(t->doc(), QLatin1String("metadata"));
+    metadata.setAttribute(QLatin1String("xmlns"), XMPP_MAM_NAMESPACE);
+    iq.appendChild(metadata);
+    iq.appendChild(makeMAMFilter().toXml(t->doc()));
+
+    t->send(iq);
+}
+
+// Note: Set `j` to a resource if you just want to query that resource
+// if you want to query all resources, set `j` to the bare JID
+
+// Filter by time range
+void MAMTask::get(const Jid &j, const QDateTime &from, const QDateTime &to, const bool allowMUCArchives,
+                  int mamPageSize, int mamMaxMessages, bool flipPages, bool backwards)
+{
+    d->archive         = {};
+    d->messagesFetched = 0;
+    d->metadataFetched = false;
+
+    d->j                = j;
+    d->from             = from;
+    d->to               = to;
+    d->allowMUCArchives = allowMUCArchives;
+    d->mamPageSize      = mamPageSize;
+    d->mamMaxMessages   = mamMaxMessages;
+    d->flipPages        = flipPages;
+    d->backwards        = backwards;
+}
+
+// Filter by id range
+void MAMTask::get(const Jid &j, const QString &from_id, const QString &to_id, const bool allowMUCArchives,
+                  int mamPageSize, int mamMaxMessages, bool flipPages, bool backwards)
+{
+    d->archive         = {};
+    d->messagesFetched = 0;
+    d->metadataFetched = false;
+
+    d->j                = j;
+    d->from_id          = from_id;
+    d->to_id            = to_id;
+    d->allowMUCArchives = allowMUCArchives;
+    d->mamPageSize      = mamPageSize;
+    d->mamMaxMessages   = mamMaxMessages;
+    d->flipPages        = flipPages;
+    d->backwards        = backwards;
+}
+
+void MAMTask::onGo() { d->getArchiveMetadata(this); }
+
+bool MAMTask::take(const QDomElement &x)
+{
+    if (d->metadataFetched) {
+        if (iqVerify(x, QString(), id())) {
+            if (!x.elementsByTagNameNS(QLatin1String("urn:ietf:params:xml:ns:xmpp-stanzas"),
+                                       QLatin1String("item-not-found"))
+                     .isEmpty()) {
+                setError(2, "First or last stanza UID of filter was not found in the archive");
+                return true;
+            } else if (!x.elementsByTagNameNS(XMPP_MAM_NAMESPACE, QLatin1String("fin")).isEmpty()) {
+                // We are done?
+                setSuccess();
+                return true;
+            }
+            // Probably ignore it
+            return false;
+        }
+
+        d->archive.append(x);
+        d->lastArchiveID   = x.attribute(QLatin1String("id"));
+        d->messagesFetched = d->messagesFetched + 1;
+
+        // Check if we are done
+        if (x.attribute(QLatin1String("id")) == d->lastID || d->messagesFetched >= d->mamMaxMessages) {
+            setSuccess();
+        } else if (d->messagesFetched % d->mamPageSize == 0) {
+            d->getPage(this);
+        }
+    } else {
+        if (!iqVerify(x, QString(), id()) || x.elementsByTagName(QLatin1String("metadata")).isEmpty())
```

Comment:
I think it worth moving metadata part to a separate task and initiate it somewhere on client login from manager. 
If there are multiple simultaneous archive queries while metadata is not completed yet, those can be queued somewhere in the manager (for example `onGo` can put in a queue instead of `send()`)

Reply:
I'd say it should be kept this way until we can prove it works